snowfall/training/asg_graph.py

```diff
@@ -0,0 +1,128 @@
+# Copyright (c)  2020  Xiaomi Corp.       (author: Fangjun Kuang)
+
+from functools import lru_cache
+from typing import Iterable
+from typing import List
+
+import re
+
+import k2
+import torch
+
+from .ctc_graph import build_ctc_topo
+
+
+def filter_disambig_symbols(symbol_table: k2.SymbolTable,
+                            pattern: str = '#') -> List[int]:
+    '''Return a list of phone IDs containing no disambiguation symbols.
+
+    Caution:
+      You may need to remove the ID with value 0 if the return value
+      is used to create a phone LM.
+
+    Args:
+      symbol_table:
+        A symbol table in k2.
+      pattern:
+        Symbols containing this pattern are disambiguation symbols.
+    Returns:
+      Return a list of symbol IDs excluding those from disambiguation symbols.
+    '''
+    regex = re.compile(pattern)
+    symbols = symbol_table.symbols
+    ans = []
+    for s in symbols:
+        if not regex.search(s):
+            ans.append(symbol_table[s])
+    return ans
+
+
+def create_bigram_phone_lm(phones: List[int]) -> k2.Fsa:
+    '''Create a bigram phone LM.
+    The resulting FSA (P) has a start-state and a state for
+    each phone 0, 1, ....; and each of the above-mentioned states
+    has a transition to the state for each phone and also to the final-state.
+
+    Caution:
+      blank is not a phone.
+    '''
+    final_state = len(phones) + 1
+    rules = ''
+    for i in range(1, final_state):
+        rules += f'0 {i} {phones[i-1]} 0.0\n'
+
+    for i in range(1, final_state):
+        for j in range(1, final_state):
+            rules += f'{i} {j} {phones[j-1]} 0.0\n'
+        rules += f'{i} {final_state} -1 0.0\n'
+    rules += f'{final_state}'
+    print(len(rules.split('\n')))
+    return k2.Fsa.from_str(rules)
+
+
+class AsgTrainingGraphCompiler(object):
+
+    def __init__(self,
+                 L_inv: k2.Fsa,
+                 phones: k2.SymbolTable,
+                 words: k2.SymbolTable,
+                 oov: str = '<UNK>'):
+        '''
+        Args:
+          L_inv:
+            Its labels are words, while its aux_labels are phones.
+        phones:
+          The phone symbol table.
+        words:
+          The word symbol table.
+        oov:
+          Out of vocabulary word.
+        '''
+
+        if L_inv.properties & k2.fsa_properties.ARC_SORTED != 0:
+            L_inv = k2.arc_sort(L_inv)
+
+        assert L_inv.requires_grad is False
+
+        assert oov in words
+
+        self.L_inv = L_inv
+        self.phones = phones
+        self.words = words
+        self.oov = oov
+
+        # note that `non_disambig_ids` contains 0, which represents
+        # blank in the context of CTC.
+        non_disambig_ids = filter_disambig_symbols(phones)
+
+        ctc_topo = build_ctc_topo(non_disambig_ids)
+        assert ctc_topo.requires_grad is False
+
+        self.ctc_topo = ctc_topo
+
+    def compile(self, texts: Iterable[str], P: k2.Fsa) -> k2.Fsa:
```

Comment:
@danpovey 
Could you help check that the graph building process is correct?

The next step I am going to do is:

-  den = intersect_dense(dense_fsa, den)
-  den_tot_scores = get_tot_scores(den) # in log semiring

-  num = intersect_dense(dense_fsa, num)
-  num_tot_scores = get_tot_scores(num) # in log semiring

(-1 * (num_tot_scores - den_tot_scores)).sum().backward()

Reply:
Remember we need to do .detach() on the version of `den` that we pass to intersect_dense(); and we need to, at some point, compose the num graphs with `den` [keeping the derivatives] so we can use that to train the scores in `den`, which we can make part of the model; we need set_scores_normalized(), if we implemented that already.
That's one possibility; another is to have the scores un-normalized and include the deriv contributions from both `num` and `den`, like I think they do in gtn; but this is less ideal as the model will get used to an LM that may not act like a 'real' LM (i.e. maximum-likelihood estimated).